pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/ReplicatedSubscriptionSnapshotCache.java

```diff
@@ -33,29 +35,96 @@
 public class ReplicatedSubscriptionSnapshotCache {
     private final String subscription;
     private final NavigableMap<Position, ReplicatedSubscriptionsSnapshot> snapshots;
+    // Used to record the timestamp of snapshots location, which will be used to adjust cache update frequency later.
+    private final NavigableMap<Position, Long> positionToTimestamp;
     private final int maxSnapshotToCache;
+    private final int snapshotFrequencyMillis;
 
-    public ReplicatedSubscriptionSnapshotCache(String subscription, int maxSnapshotToCache) {
+    public ReplicatedSubscriptionSnapshotCache(String subscription, int maxSnapshotToCache,
+                                               int snapshotFrequencyMillis) {
         this.subscription = subscription;
         this.snapshots = new TreeMap<>();
+        this.positionToTimestamp = new TreeMap<>();
         this.maxSnapshotToCache = maxSnapshotToCache;
+        this.snapshotFrequencyMillis = snapshotFrequencyMillis;
     }
 
-    public synchronized void addNewSnapshot(ReplicatedSubscriptionsSnapshot snapshot) {
+    public synchronized void addNewSnapshot(ReplicatedSubscriptionsSnapshot snapshot, long publishTime) {
```

Comment:
Please add a javadoc comment to this method and explain that `publishTime` is the `publishTime` of the `ReplicatedSubscriptionsSnapshot` marker message (entry). Since the `publishTime` originates from the same broker or at least from the same cluster, it's assumed that the time is reliable and clock skew is not a problem.